xarray/coding/times.py

```diff
@@ -1509,8 +1509,11 @@ def decode(self, variable: Variable, name: T_Name = None) -> Variable:
                         f"present in encoding. Check the encoding parameters "
                         f"of variable {name!r}."
                     )
-                dtype = pop_to(attrs, encoding, "dtype", name=name)
-                dtype = np.dtype(dtype)
+                # Overwrite the on-disk dtype encoding, which is numeric, with
+                # the dtype attribute stored on disk, which corresponds to
+                # a timedelta64 dtype.
+                encoding["dtype"] = attrs.pop("dtype")
```

Comment:
I don't think we want to set `datetime64` here as the encoded dtype, because this information is preserved in the dtype of the resulting `Variable`. We should leave `encoding['dtype']` as the raw numeric dtype on disk.

Reply:
I went back and forth on this a bit in the initial PR.  Would we be open to renaming the attribute that encodes the `timedelta64` dtype `"xarray_dtype"`?  This way we would not need to worry about overwriting the numeric dtype in the `encoding` and can still retain information about the `timedelta64` encoding method used.

I think it is potentially useful, at least while we still have two different methods of decoding `timedelta64` values, to be able to distinguish between them for round-tripping purposes (i.e. so that we do not add new dtype attributes to files that did not have them previously).  Maybe that should not be a priority though.

Reply:
> I think it is potentially useful, at least while we still have two different methods of decoding `timedelta64` values, to be able to distinguish between them for round-tripping purposes (i.e. so that we do not add new dtype attributes to files that did not have them previously). Maybe that should not be a priority though.

I think users probably _do_ want new `dtype` attributes saved if they are re-writing data with Xarray. This makes their datasets forward compatible with future versions of Xarray, and silences the deprecation warning about decoding changing in the future, in a way that feels cleaner to me than adding `decode_timedelta=True` to every `open_dataset()` call.

Reply:
Ok great, that should simplify things.